### PR TITLE
QCArchive dataset curation

### PR DESCRIPTION
**Pull request by maxentile**
Refining utilities from @yuanqing-wang ([utils.qc_archive_utils](https://github.com/choderalab/espaloma/blob/f595317de598eb5e7a556a909a71b375defdbf57/espaloma/data/qcarchive_utils.py), and [scripts/qca](https://github.com/choderalab/espaloma/tree/ee2cd77d36501b09ab052051d0df9500e0821636/scripts/qca)), in this PR we'll work towards fetching and curating datasets from QCArchive.

- [x] Select datasets, closely following OpenFF 1.2 training dataset selection , as described here https://zenodo.org/record/3777278 . (Possibly also including "OpenFF Protein Fragments v1.0", "OpenFF Protein Fragments v2.0", and other peptide datasets in development.)
- [x] Download selected datasets from QCArchive, with appropriate error handling
- [x] Save molecules, trajectories, energies, and gradients in standard file formats 
- [x] Define and apply any dataset filtering or weighting necessary

#### Comments on formats

Currently, these are downloaded and saved as `espaloma.Graph` datasets in PyTorch `.th` files.

Requirements for update:
* retain necessary provenance
* save in a format which avoids dependence on which `espaloma` commit was used (dataset needed to be regenerated when we added impropers, because the graph object not containing impropers were incompatible with the updated version)
* provide a script to convert the saved dataset to an `esp.Graph` dataset

Intended solution: save the dataset in a pandas dataframe hdf5 file containing record names, offmols, xyz trajectories, energies, gradients, and all provenance returned by QCA at the `collection` level (but not at the per-snapshot level).

Full provenance at the per-snapshot level can later be recovered by querying QCArchive using the stored record names and snapshot indices.

#### Comments on filtering / weighting

Currently, the fetched `OptimizationDataset` is filtered to retain species that have at least 1000 configuration snapshots, but it's unclear what this filter might also select for (e.g. this probably selects for molecules where the optimizer doesn't reach its convergence criteria quickly).

Possibly we want to filter by other criteria (e.g. to discard snapshots with relative energy exceeding some threshold -- a quick inspection shows some snapshots with relative energy as high as ~6500 kcal/mol (see `'OpenFF Full Optimization Benchmark 1'` dataset record `'c[c@h](cn1cnc2c1ncnc2n)ocp(=o)([o-])[o-]-5'`), which we probably do not want to include in our training.

Some random example traces from `'OpenFF Full Optimization Benchmark 1'` dataset: 

image
image
image
image

One approach would be to discard snapshots whose relative energy exceeds a threshold like `(5 kcal/mol) * (num_atoms)` or something.

An alternative to hard thresholding, suggested by @yuanqing-wang (and which I think is used in Force Balance) is to apply a soft weighting function to emphasize lower-energy snapshots, such as `exp[ - beta (u - u0) ]` where `u0` is the energy of the presumed global minimum and `beta` is a inverse-temperature-y coefficient saying how sharply to penalize excursions from `u0`.

### CODE REVIEW

**Comment by jchodera** [comment]
> Currently, the fetched OptimizationDataset is filtered to retain species that have at least 1000 configuration snapshots, but it's unclear what this filter might also select for (e.g. this probably selects for molecules where the optimizer doesn't reach its convergence criteria quickly).

@maxentile : Is that 1000 conformations, or 1000 snapshots total (among all pooled gradient descent snapshots)?

I wonder if this doesn't filter out simple molecules with few conformers (e.g. benzene) that just make life harder in training a model.

Is there a rationale for having a filter like this, other than the need to subtract an additive offset for each molecule?

**Comment by maxentile** [comment]
>@maxentile : Is that 1000 conformations, or 1000 snapshots total (among all pooled gradient descent snapshots)?

1000 snapshots total, I believe.

>I wonder if this doesn't filter out simple molecules with few conformers (e.g. benzene) that just make life harder in training a model.

I would suspect this also -- part of my concern is I don't know what else a filter like this might inadvertently select for.

>Is there a rationale for having a filter like this, other than the need to subtract an additive offset for each molecule?

I think part of the rationale was related to efficiency of batching (perhaps @yuanqing-wang can describe a bit more).

One intent of this PR is to switch to the Gen 2 datasets used in Parsley fitting, and to define any alternative filters if necessary.

* For the Gen 2 TorsionDriveDatasets, I don't see any reason for filtering to be required -- these are all very carefully selected and highly informative.
* For the Gen 2 OptimizationDatasets, I think we may want to filter out any really high-energy snapshots, if present. I noticed some extremely high-energy snapshots in a few trajectories of the `'OpenFF Full Optimization Benchmark 1'` (up to ~6500 kcal/mol). One other thing we might flag for removal is if the molecule connectivity changed during optimization (e.g. in [Fig 1 of the benchmarkff preprint](https://chemrxiv.org/articles/preprint/Benchmark_Assessment_of_Molecular_Geometries_and_Energies_from_Small_Molecule_Force_Fields/12551867)). Checking if this behavior is present in any of the Gen 2 OptimizationDatasets...

**Comment by jchodera** [comment]
> For the Gen 2 OptimizationDatasets, I think we may want to filter out any really high-energy snapshots, if present. I noticed some extremely high-energy snapshots in a few trajectories of the 'OpenFF Full Optimization Benchmark 1' (up to ~6500 kcal/mol). One other thing we might flag for removal is if the molecule connectivity changed during optimization (e.g. in Fig 1 of the benchmarkff preprint). Checking if this behavior is present in any of the Gen 2 OptimizationDatasets...

Agreed that these are sensible filters, especially if we're not down-weighting high-energy snapshots!

**Comment by maxentile** [comment]
This is nearly ready for review -- still need to check on energy traces to see if we want to apply a filter at this stage.

@yuanqing-wang  -- I could use a hand with a conversion script from pandas h5 to espaloma graph dataset -- here's [a short notebook](https://github.com/choderalab/espaloma/blob/524ae452b588846ac08946bb8ab977dd9684897a/scripts/qca/merge%20by%20offmol.ipynb) with one way to merge records by offmol.

**Comment by maxentile** [comment]
Just noticed something fishy.

tl;dr -- toolkit-dependent behavior of Openforcefield `Molecule.from_qcschema(entry)` for least 2 molecules in fetched datasets

When attempting to analyze the max relative energy per molecule on my laptop, I encountered an OpenEye stereochem errors. This is triggered, for example, when calling `offmol.to_smiles()` on this offmol that was downloaded on `lilac` and saved in `Coverage.h5`:
image


However, when I attempted to reproduce this locally by fetching that offmol for the record in the same way from my laptop, I got a subtly different molecule that doesn't have this error...

image


Further, `offmol_fetched_on_laptop.to_dict() != offmol_fetched_on_lilac.to_dict()`.

This suggests that `Molecule.from_qcschema(ds.get_entry(record_name))` had environment-dependent behavior for at least two `record_name`s.

In my laptop environment, I'm using OpenEye 2020.1.0 and Openforcefield 0.7.2.

In the lilac environment where QCA was queried and this dataset was formed, I had used RDKit 2020.03.6 and Openforcefield 0.7.1.

This issue affects only the records `'c1[c@@h]2[c@@h]([c@h]2o)c[nh2+]1-0'` and `'c1c[nh+]c(cc1cl)[s@@](=o)[o-]-0'` in OptimizationDataset "OpenFF Gen 2 Opt Set 2 Coverage".

Otherwise, `offmol.to_smiles()` works with OpenEye backend for all remaining records in all other datasets fetched.

Options:
1. Discard these two molecules from the currently fetched dataset, possibly returning to the issue later
2. Fetch all molecule entries from QCArchive again, but ensuring Openforcefield is using the latest OpenEye rather than RDKit backend when doing the `Molecule.from_qcschema(entry)` conversion
3. Look at this further, if this indicates something deeper amiss

I'd vote for option 1.

**Comment by maxentile** [comment]
>toolkit-dependent behavior of Openforcefield Molecule.from_qcschema(entry) for least 2 molecules in fetched datasets

Revise that from "at least 2" to "exactly 2" -- ran a [quick sanity-check](https://github.com/choderalab/espaloma/blob/cb6ad031cbcf584d84af594962ea728ac6be727d/scripts/qca/check%20for%20toolkit-dependent%20interpretation%20of%20fetched%20molecules.ipynb) that this issue is not more pervasive, strengthening my vote to discard these two molecules.

**Review by yuanqing-wang — APPROVED** [review]

**Comment by maxentile** [comment]
@yuanqing-wang : Nice! The conversions to `esp.data.dataset.GraphDataset` look safe to me, and good call moving the datasets from `espaloma/data` to `data/`. The `__add__` method for datasets is also a nice touch, although I'm not sure what the behavior of `graphs + graphs` is.

In a moment, I'll update the dataset paths in the notebooks to match.

**Comment by yuanqing-wang** [comment]
> although I'm not sure what the behavior of `graphs + graphs` is.

Sorry I didn't clarify---`Dataset.graphs` is just a list

**Comment by yuanqing-wang** [comment]
also @maxentile could you double check that there's no data in the actual git? the 15k lines increase in this PR includes the notebook but not the data right?

**Comment by maxentile** [comment]
>also @maxentile could you double check that there's no data in the actual git? the 15k lines increase in this PR includes the notebook but not the data right?

I believe the `.h5` files are successfully tracked by `git-lfs` and contribute just ~2 lines each to the git history.

The bulk of these lines are the notebooks and the lists of batched QCA queries, which I saved to disk to be on the safe side in case there was any evidence of non-deterministic behavior in the bulk QCA query. The text containing the exact QCA queries used can safely be deleted since they are now stored in each `df.index` (and can be regenerated by `make_batches.py` if needed). I'll delete these txt files in a moment, unless there's a reason to hold on to them separately.

**Comment by maxentile** [comment]
Overall dataset filtering proposed:
* Apply relative energy threshold of 100 kcal/mol
   * option 1: discard any _snapshot_ exceeding threshold -- this would discard ~0.01% of the snapshots in the eMolecules dataset (and keep all snapshots in all other datasets)
   * option 2: discard any _molecule_ which ever encountered a conformer exceeding threshold -- this would discard ~1.9% of molecules in the eMolecules dataset (and keep all molecules in all other datasets)
* Discard potentially ambiguous topologies
   * discard two records, `'c1[c@@h]2[c@@h]([c@h]2o)c[nh2+]1-0'` and `'c1c[nh+]c(cc1cl)[s@@](=o)[o-]-0'` from OptimizationDataset "OpenFF Gen 2 Opt Set 2 Coverage", for reasons mentioned in https://github.com/choderalab/espaloma/pull/47#issuecomment-713753019

@yuanqing-wang : are there other filters we should consider applying at this stage?

**Comment by yuanqing-wang** [comment]
I couldn't think of any. We'll exam the outliers as we move on.

**Comment by maxentile** [comment]
Reported the topology ambiguity encountered at https://github.com/choderalab/espaloma/pull/47#issuecomment-713753019 to OpenForcefield in this issue: https://github.com/openforcefield/openforcefield/issues/749
